### PR TITLE
[Dashboard]: Convert mappings to a map for better performance

### PR DESCRIPTION
**Pull request by uladzimirdev**
Implementing @oleggromov idea https://github.com/metabase/metabase/pull/38289#discussion_r1472653468

Access is O(1) because of using map

### CODE REVIEW

File: frontend/src/metabase/parameters/utils/dashboards.ts [hunk 1]
```diff
@@ -109,17 +109,41 @@ function getMappings(dashcards: DashboardCard[]): ExtendedMapping[] {
...
+  for (const mapping of mappings) {
+    if (!mappingsByParameterId.has(mapping.parameter_id)) {
+      mappingsByParameterId.set(mapping.parameter_id, new Map());
+    }
+
+    const targetKey = Symbol.for(mapping.target.toString());
+    if (!mappingsByParameterId.get(mapping.parameter_id)?.has(targetKey)) {
+      mappingsByParameterId.get(mapping.parameter_id)?.set(targetKey, mapping);
+    }
+  }
```
Comment (oleggromov) [inline frontend/src/metabase/parameters/utils/dashboards.ts:129]: Nitpick: it appears to me that this is unnecessarily verbose.
I would suggest using a plan object, avoid symbols and find a way of doing something in the spirit of `assocIn(mappingsByParameterId, [mapping.parameter_id, mapping.target.toString()], mapping)` within the loop.

It's a matter of taste though 😉

**Review by oleggromov — APPROVED** [review]
Left a tiny suggestion, but otherwise LGTM!

File: frontend/src/metabase/parameters/utils/dashboards.ts (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/parameters/utils/dashboards.ts:129]: +1, I'd also prefer a plain object unless `Map` provides some significant performance win of something

File: frontend/src/metabase/parameters/utils/dashboards.ts [hunk 2]
```diff
@@ -109,17 +109,41 @@ function getMappings(dashcards: DashboardCard[]): ExtendedMapping[] {
...
 ): UiParameter[] {
   const { parameters, dashcards } = dashboard;
-  const mappings = getMappings(dashcards as DashboardCard[]);
+  const mappings = getMappingsByParameterId(dashcards as DashboardCard[]);
   const uiParameters: UiParameter[] = (parameters || []).map(parameter => {
     if (isFieldFilterParameter(parameter)) {
-      return buildFieldFilterUiParameter(parameter, mappings, metadata);
+      return buildFieldFilterUiParameter(
+        parameter,
+        [...(mappings.get(parameter.id)?.values() ?? [])],
```
Comment (kulyk) [inline frontend/src/metabase/parameters/utils/dashboards.ts:144]: nit: there are many things going on here, and it feels a bit hard to parse
might be worth unpacking it into several lines

**Review by kulyk — APPROVED** [review]
LGTM, just two code-style comments

**Comment by uladzimirdev** [comment]
the main problem was resolved in a different way, this PR doesn't change a lot, so closing
